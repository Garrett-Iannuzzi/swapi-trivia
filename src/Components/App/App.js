import React, { Component } from 'react';
import './App.scss';
import Nav from '../Nav/Nav.js';
import Login from '../Login/Login.js';
import MovieContainer from '../MovieContainer/MovieContainer.js';
import Error from '../Error/Error.js';
import CharacterContainer from '../CharacterContainer/CharacterContainer.js';
import { Route } from 'react-router-dom';



class App extends Component {
  constructor() {
    super()
    this.state = {
      display: [],
      characters: [],
      userInfo: {},
      isLoading: true,
      hasError: true
    }
  }

  componentDidMount = () => {
    fetch('https://swapi.co/api/films/')
      .then(res => res.json())
      .then(data => this.setState({ display: data.results }))
      .then(data => this.setState({ isLoading: false }))
  }

  handleLoginError = (loginStatus) => {
    this.setState({ hasError: loginStatus });
  }

  resetUserInfo = () => {
    this.setState({  hasError: true, userInfo: {}, characters: [], })
  }

  handleUserInfo = (info) => {
    this.setState({ userInfo: info })
  }

  resetCharacters = () => {
    this.setState({ characters: [] })
  }

  getCharacterData = (id) => {
    this.setState({ isLoading: true })
    fetch('https://swapi.co/api/films/')
    .then(res => res.json())
    .then(data => this.handleCharacterFetch(data, id))
    .then(characterData => this.setState({ characters: characterData, isLoading: false }))
  }

  handleCharacterFetch = (movies, id) => {
    let movieData = movies.results.filter(movie => movie.episode_id === Number(id));
    let firstTenCharacters = movieData[0].characters.slice(0, 10);
    const promises = firstTenCharacters.map(character => {
      return fetch(character)
      .then(res => res.json())
      .then(data => this.handleCharacterInfo(data))
      .then(data => {
        let object1 = data[0];
        let object2 = data[1];
        let array = data[2];
        let infoObject = {
          name: object1.name,
          population: object1.population,
          species: object2.species,
          creature: object2.creature,
          relatedFilms: array,
          character: object2.character,
          openingCrawl: movieData[0].opening_crawl
        }
        return infoObject;
      })
      .catch(err => console.log(err));
    });
    return Promise.all(promises);
  }

  handleCharacterInfo = (character) => {
    const homeWorldInfo = this.handleCharacterHomeworld(character);
    const speciesInfo = this.handleCharacterSpecies(character);
    const relatedFilmsData = this.handleCharacterRelatedFilm(character);
    let filmsPromises = Promise.all(relatedFilmsData)
    let promises = [homeWorldInfo, speciesInfo, filmsPromises];
    return Promise.all(promises);
  }

  handleCharacterHomeworld = (character) => {
    return fetch(character.homeworld)
     .then(res => res.json())
     .then(data => ({
      name: data.name,
      population: data.population
      }))
      .catch(err => console.log(err));
  }

  handleCharacterSpecies = (character) => {
    return fetch(character.species)
    .then(res => res.json())
    .then(data => ({
      species: data.name,
      creature: data.classification,
      character: character.name
    }))
    .catch(err => console.log(err));
  }

  handleCharacterRelatedFilm = (character) => {
    return character.films.map(film => {
      return fetch(film)
      .then(res => res.json())
      .then(data => ({relatedFilms: data.title}))
      .catch(err => console.log(err));
    });
  }

  render() {
    return (
      <div className='page-container'>
          <Route path='/movies' render={ () =>
            !this.state.hasError &&
              <Nav
                userInfo={this.state.userInfo}
                handleLoginError={this.handleLoginError}
                resetUserInfo={this.resetUserInfo}
              />
            }/>
          <Route exact path='/' render={ () =>
              <Login
                handleLoginError={this.handleLoginError}
                handleUserInfo={this.handleUserInfo}
              />
            }/>
          <Route exact path='/movies' render={ () =>
            {if (this.state.hasError) {
              return <Error />
             } else if (this.state.isLoading ) {
              return <h2>Loading...</h2>
            } else {
             return <MovieContainer
               getCharacterData={this.getCharacterData}
               movieCards={this.state.display}
             />
           }}
          }/>
          <Route path='/movies/:movies_id' render={ () => {
            return(
              this.state.characters.length ?
               <CharacterContainer
                characters={this.state.characters}
                resetCharacters={this.resetCharacters}
              /> :
              <h2>Loading...</h2>
            )
          }} />
      </div>
    );
  }
}

export default App;
