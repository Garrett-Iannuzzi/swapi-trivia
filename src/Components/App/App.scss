.page-container {
  height: 100vh;
  background-image: url(https://external-preview.redd.it/4gQ9nVb85g1MFpaxX4D7M641Lx73OEuhbqbSQfLpYU8.jpg?auto=webp&s=f074a9b622a99d327a25d19213e4a299253ba0e7);
  background-size: cover;
  display: flex;
  flex-direction: column;
  align-items: center;
}

main {
  width: 100%;
  height: 100vh;
  display: flex;
  align-items: center;
}
