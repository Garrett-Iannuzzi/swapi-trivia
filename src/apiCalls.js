
export const getAllMovies = () => {
  return fetch('https://swapi.co/api/films/')
  .then(response => {
    if(!response.ok) {
      throw Error('Error fetchings movies')
    }
    return response.json()
  })
}

export const getCharacterHomeworld = (character) => {
  return fetch(character.homeworld)
  .then(res => res.json())
  .then(data => ({
    name: data.name,
    population: data.population
    }))
}

export const getCharacterSpecies = (character) => {
  return fetch(character.species)
  .then(res => res.json())
  .then(data => ({
    species: data.name,
    creature: data.classification,
    character: character.name
  }))
}

export const getCharacterRelatedFilm = (film) => {
  return fetch(film)
  .then(res => res.json())
  .then(data => ({relatedFilms: data.title}))
}