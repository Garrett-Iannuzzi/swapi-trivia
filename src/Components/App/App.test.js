import React from 'react';
import ReactDOM from 'react-dom';
import { shallow } from 'enzyme';
import App from '../App/App.js';
import { getAllMovies, getCharacterHomeworld, getCharacterSpecies, getCharacterRelatedFilm } from '../../apiCalls.js';

jest.mock('../../apiCalls.js');

describe('App', () => {

  let wrapper, mockFilms;

  beforeEach(() => {
    mockFilms = [{
      title: 'A new Hope',
      episode_id: 1,
      release_date: '1983/4/12',
      characters: [{name: 'Luke', homeworld: 'url', species: 'url', relatedFilms: ['url', 'url']}],
      opening_crawl: 'In the beginning...'
    }]

    getAllMovies.mockImplementation(() => {
      return Promise.resolve(mockFilms)
    });

    wrapper = shallow(<App />);

  });

  it('Should match the snapshot with all data', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should call getAllMovies after component is mounted', () => {
    shallow(<App />);

    expect(getAllMovies).toHaveBeenCalled();
  });

  it('handleCharacterHomeworld should call getCharacterHomeworld when invoked' , () => {
    let mockHomeworldData = {name: 'Tatooine', population: '200000'}

    getCharacterHomeworld.mockImplementation(() => {
      return Promise.resolve(mockHomeworldData);
    });

    const wrapper = shallow(<App />);
    wrapper.instance().handleCharacterHomeworld();
    expect(getCharacterHomeworld).toHaveBeenCalled();
  });

  it('handleCharacterHomeworld should return the correct homeworld data' , () => {
    let mockHomeworldData = {name: 'Tatooine', population: '200000'}

    getCharacterHomeworld.mockImplementation(() => {
      return Promise.resolve(mockHomeworldData);
    });

    const wrapper = shallow(<App />);
    expect(wrapper.instance().handleCharacterHomeworld()).resolves.toEqual(mockHomeworldData);
  });

  it('handleCharacterSpecies should call getCharacterSpecies when invoked' , () => {
    let mockSpeciesData = {species: 'human', creature: 'mammal', character: 'john'}

    getCharacterSpecies.mockImplementation(() => {
      return Promise.resolve(mockSpeciesData);
    });

    const wrapper = shallow(<App />);
    wrapper.instance().handleCharacterSpecies();
    expect(getCharacterSpecies).toHaveBeenCalled();
  });

  it('handleCharacterSpecies should return the correct species data' , () => {
    let mockSpeciesData = {species: 'human', creature: 'mammal', character: 'john'}

    getCharacterSpecies.mockImplementation(() => {
      return Promise.resolve(mockSpeciesData);
    });

    const wrapper = shallow(<App />);
    expect(wrapper.instance().handleCharacterSpecies()).resolves.toEqual(mockSpeciesData);
  });

  it('handleCharacterRelatedFilm should call getCharacterRelatedFilm when invoked' , () => {
    let mockRelatedFilmData = {relatedFilms: 'A new hope'}
    let mockCharacter = {films: ['A new hope', 'The last Jedi']};

    getCharacterRelatedFilm.mockImplementation(() => {
      return Promise.resolve(mockRelatedFilmData);
    });

    const wrapper = shallow(<App />);
    wrapper.instance().handleCharacterRelatedFilm(mockCharacter);
    expect(getCharacterRelatedFilm).toHaveBeenCalled();
  });

  it('handleCharacterRelatedFilm should return the correct related film data' , () => {
    let mockRelatedFilmData = {relatedFilms: 'A new hope'};
    let mockCharacter = {films: ['A new hope', 'The last Jedi']};
    let mockReturnedRelatedFilmData = Promise.resolve([{relatedFilms: 'A new hope'}, {relatedFilms: 'A new hope'}]);

    getCharacterRelatedFilm.mockImplementation(() => {
      return Promise.resolve(mockRelatedFilmData);
    });

    const wrapper = shallow(<App />);
    expect(Promise.resolve(wrapper.instance().handleCharacterRelatedFilm(mockCharacter))).toEqual(mockReturnedRelatedFilmData);
  });

  it('Should update state with a new hasError when handleLoginError is called', () => {
    wrapper.setState({hasError: true});
    wrapper.instance().handleLoginError(false);

    expect(wrapper.state('hasError')).toEqual(false);
  });

  it('Should update state with a new hasError, userInfo and characters when resetUserInfo is called', () => {
    wrapper.setState({
      hasError: false,
      userInfo: {
        name: 'Luke Skywalker',
        quote: 'You are not my dad!!!',
        skillLevel: 'Jedi Master'
      },
      characters: [
        {
          character: "C-3PO",
          creature: "artificial",
          name: "Tatooine",
          openingCrawl: "Turmoil has engulfed th",
          population: "200000",
          relatedFilms: [{
            relatedFilms: "The Empire Strikes Back"},
            {relatedFilms: "Attack of the Clones"}],
          species: "Droid"
        },
        {
          character: "Jabba Desilijic Tiure",
          creature: "gastropod",
          name: "Nal Hutta",
          openingCrawl: "Turmoil has engulfed th",
          population: "7000000000",
          relatedFilms: [{
            relatedFilms: "The Phantom Menace"},
            {relatedFilms: "Return of the Jedi"}],
          species: "Hutt"
        }
      ]
    });

    wrapper.instance().resetUserInfo();
    expect(wrapper.state('hasError')).toEqual(true);
    expect(wrapper.state('userInfo')).toEqual({});
    expect(wrapper.state('characters')).toEqual([])
  });

  it('Should update state with userInfo when handleUserInfo is called', () => {
    const info = {
      name: 'Luke Skywalker',
      quote: 'Strong you are',
      skillLevel: 'Jedi Master'
    }

    wrapper.setState({userInfo: {}});
    wrapper.instance().handleUserInfo(info);

    expect(wrapper.state('userInfo')).toEqual(info);
  });

  it('Should reset characters state when resetCharacters is called', () => {
    wrapper.setState({characters: [
      {
        character: "C-3PO",
        creature: "artificial",
        name: "Tatooine",
        openingCrawl: "Turmoil has engulfed th",
        population: "200000",
        relatedFilms: [{
          relatedFilms: "The Empire Strikes Back"},
          {relatedFilms: "Attack of the Clones"}],
        species: "Droid"
      },
      {
        character: "Jabba Desilijic Tiure",
        creature: "gastropod",
        name: "Nal Hutta",
        openingCrawl: "Turmoil has engulfed th",
        population: "7000000000",
        relatedFilms: [{
          relatedFilms: "The Phantom Menace"},
          {relatedFilms: "Return of the Jedi"}],
        species: "Hutt"
      }
    ]});

    wrapper.instance().resetCharacters();

    expect(wrapper.state('characters')).toEqual([]);
  });

  it('Should add selected favorite to favorites when handleFavorites is called', () => {
    const mockFavorite = {
      character: "C-3PO",
      creature: "artificial",
      name: "Tatooine",
      population: "200000",
      relatedFilms: [
        {relatedFilms: "The Empire Strikes Back"},
        {relatedFilms: "Attack of the Clones"}
      ],
      species: "Droid"
    }

    wrapper.setState({ favorites: [] });

    wrapper.instance().handleFavorites(mockFavorite);
    expect(wrapper.state('favorites')).toEqual([mockFavorite]);
  });

  it('Should remove selectecd favorite if it is already in favorite when handleFavorites is called', () => {
    const mockFavorite2 = {
      character: "C-3PO",
      creature: "artificial",
      name: "Tatooine",
      population: "200000",
      relatedFilms: [
        {relatedFilms: "The Empire Strikes Back"},
        {relatedFilms: "Attack of the Clones"}
      ],
      species: "Droid"
    }

    wrapper.setState({ favorites: [mockFavorite2] });

    wrapper.instance().handleFavorites(mockFavorite2);
    expect(wrapper.state('favorites')).toEqual([]);
  });
});
