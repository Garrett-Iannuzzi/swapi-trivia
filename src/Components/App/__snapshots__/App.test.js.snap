// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Should match the snapshot with all data 1`] = `
<div
  className="page-container"
>
  <Route
    path="/movies"
    render={[Function]}
  />
  <Route
    exact={true}
    path="/"
    render={[Function]}
  />
  <Route
    exact={true}
    path="/movies"
    render={[Function]}
  />
  <Route
    path="/movies/:movies_id"
    render={[Function]}
  />
  <Route
    path="/favorites"
    render={[Function]}
  />
</div>
`;
